pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,721 @@
+// select and to which targets to do it.
+type AttachRule struct {
+	// LibraryNameRegex defines which libraries should be matched by this rule
+	LibraryNameRegex *regexp.Regexp
+	// Targets defines the targets to which we should attach the probes, shared libraries and/or executables
+	Targets AttachTarget
+	// ProbesSelectors defines which probes should be attached and how should we validate
+	// the attachment (e.g., whether we need all probes active or just one of them, or in a best-effort basis)
+	ProbesSelector []manager.ProbesSelector
+}
+
+func (r *AttachRule) canTarget(target AttachTarget) bool {
+	return r.Targets&target != 0
+}
+
+func (r *AttachRule) matchesLibrary(path string) bool {
+	return r.canTarget(AttachToSharedLibraries) && r.LibraryNameRegex != nil && r.LibraryNameRegex.MatchString(path)
+}
+
+func (r *AttachRule) matchesExecutable(_ string) bool {
+	return r.canTarget(AttachToExecutable)
+}
+
+// AttacherConfig defines the configuration for the attacher
+type AttacherConfig struct {
+	// Rules defines a series of rules that tell the attacher how to attach the probes
+	Rules []*AttachRule
+
+	// ScanTerminatedProcessesInterval defines the interval at which we scan for terminated processes. Set
+	// to zero to disable
+	ScanTerminatedProcessesInterval time.Duration
+
+	// ProcRoot is the root directory of the proc filesystem
+	ProcRoot string
+
+	// ExcludeTargets defines the targets that should be excluded from the attacher
+	ExcludeTargets ExcludeMode
+
+	// EbpfConfig is the configuration for the eBPF program
+	EbpfConfig *ebpf.Config
+
+	// PerformInitialScan defines if the attacher should perform an initial scan of the processes before starting the monitor
+	PerformInitialScan bool
+
+	// ProcessMonitorEventStream defines whether the process monitor is using the event stream
+	ProcessMonitorEventStream bool
+}
+
+// SetDefaults configures the AttacherConfig with default values for those fields for which the compiler
+// defaults are not enough
+func (ac *AttacherConfig) SetDefaults() {
+	if ac.ScanTerminatedProcessesInterval == 0 {
+		ac.ScanTerminatedProcessesInterval = 30 * time.Minute
+	}
+
+	if ac.ProcRoot == "" {
+		ac.ProcRoot = "/proc"
+	}
+
+	if ac.EbpfConfig == nil {
+		ac.EbpfConfig = ebpf.NewConfig()
+	}
+}
+
+// ProbeManager is an interface that defines the methods that a Manager implements,
+// so that we can replace it in tests for a mock object
+type ProbeManager interface {
+	AddHook(string, *manager.Probe) error
+	DetachHook(manager.ProbeIdentificationPair) error
+	GetProbe(manager.ProbeIdentificationPair) (*manager.Probe, bool)
+}
+
+// FileRegistry is an interface that defines the methods that a FileRegistry implements, so that we can replace it in tests for a mock object
+type FileRegistry interface {
+	Register(namespacedPath string, pid uint32, activationCB, deactivationCB func(utils.FilePath) error) error
+	Unregister(uint32) error
+	Clear()
+	GetRegisteredProcesses() map[uint32]struct{}
```

Comment:
What about "blocked processes"?

Reply:
It's not needed for this code so I didn't add it in the interface.